File: cz_cli/commands/workspace.py
"""clickzetta workspace command — manage workspaces."""

from __future__ import annotations

from pathlib import Path
from typing import Any

import click

from cz_cli import output
from cz_cli.connection import get_connection
from cz_cli.logger import log_operation

try:
    import tomllib
except ImportError:
    import tomli as tomllib  # type: ignore


_PROFILES_DIR = Path.home() / ".clickzetta"
_PROFILES_FILE = _PROFILES_DIR / "profiles.toml"


@click.group("workspace")
@click.pass_context
def workspace_cmd(ctx: click.Context) -> None:
    """Manage workspaces."""


@workspace_cmd.command("current")
@click.pass_context
def current_workspace(ctx: click.Context) -> None:
    """Show current workspace."""
    fmt: str = ctx.obj["format"]
    profile: str | None = ctx.obj.get("profile")

    try:
        conn = get_connection(profile=profile)
    except Exception as exc:
        log_operation("workspace current", ok=False, error_code="CONNECTION_ERROR")
        output.error("CONNECTION_ERROR", str(exc), fmt=fmt)
        return

    timer = output.Timer()
    try:
        with timer:
            cursor = conn.cursor()
            try:
                cursor.execute("SELECT current_workspace()")
                rows = cursor.fetchall()

                if rows:
                    workspace_name = list(rows[0].values())[0]
                    log_operation("workspace current", ok=True, time_ms=timer.elapsed_ms)
                    output.success({"workspace": workspace_name}, time_ms=timer.elapsed_ms, fmt=fmt)
                else:
                    log_operation("workspace current", ok=False, error_code="NO_RESULT")
                    output.error("NO_RESULT", "No workspace found", fmt=fmt)
            finally:
                cursor.close()
    except Exception as exc:
        log_operation("workspace current", ok=False, error_code="SQL_ERROR")
        output.error("SQL_ERROR", str(exc), fmt=fmt)
    finally:
        conn.close()


@workspace_cmd.command("use")
@click.argument("name")
@click.option("--schema", default=None, help="Schema to use (defaults to current profile schema)")
@click.option("--persist", is_flag=True, help="Update profile configuration to persist the change")
@click.pass_context
def use_workspace(ctx: click.Context, name: str, schema: str | None, persist: bool) -> None:
    """Switch to a workspace using SDK hints.

    This command uses the SDK hint 'sdk.job.default.ns' to switch workspace context.
    Use --persist to also update the current profile configuration.
    """
    fmt: str = ctx.obj["format"]
    profile: str | None = ctx.obj.get("profile")

    # If persist is requested, update the profile
    if persist:
        try:
            data = _load_profiles()
            profiles = data.get("profiles", {})

            # Determine which profile to update
            profile_name = profile or "default"
            if profile_name not in profiles:
                log_operation("workspace use", ok=False, error_code="PROFILE_NOT_FOUND")
                output.error("PROFILE_NOT_FOUND", f"Profile '{profile_name}' not found", fmt=fmt)
                return

            profiles[profile_name]["workspace"] = name
            if schema:
                profiles[profile_name]["schema"] = schema

            data["profiles"] = profiles
            _save_profiles(data)

            log_operation("workspace use", ok=True)
            output.success({
                "message": f"Switched to workspace '{name}' and updated profile '{profile_name}'",
                "workspace": name,
                "schema": schema or profiles[profile_name].get("schema", "public")
            }, fmt=fmt)
        except Exception as exc:
            log_operation("workspace use", ok=False, error_code="PROFILE_UPDATE_ERROR")
            output.error("PROFILE_UPDATE_ERROR", str(exc), fmt=fmt)
    else:
        # Just show the hint that would be used
        # Note: The actual SDK hint application happens at connection/query time
        # This command primarily serves to update the profile for persistence
        schema_name = schema or "public"
        log_operation("workspace use", ok=True)
        output.success({
            "message": f"To use workspace '{name}', set SDK hint: {{'sdk.job.default.ns': '{name}.{schema_name}'}}",
            "workspace": name,
            "schema": schema_name,
            "note": "Use --persist to save this to your profile configuration"
        }, fmt=fmt)


def _load_profiles() -> dict[str, Any]:
    """Load profiles from ~/.clickzetta/profiles.toml"""
    if not _PROFILES_FILE.exists():
        return {"profiles": {}}

    try:
        with open(_PROFILES_FILE, "rb") as f:
            return tomllib.load(f)
    except Exception:
        return {"profiles": {}}


def _save_profiles(data: dict[str, Any]) -> None:
    """Save profiles to ~/.clickzetta/profiles.toml"""
    _PROFILES_DIR.mkdir(parents=True, exist_ok=True)

    # Convert to TOML format
    lines = []
    for profile_name, profile_data in data.get("profiles", {}).items():
        lines.append(f"[profiles.{profile_name}]")
        for key, value in profile_data.items():
            if isinstance(value, str):
                lines.append(f'{key} = "{value}"')
            else:
                lines.append(f"{key} = {value}")
        lines.append("")

    with open(_PROFILES_FILE, "w", encoding="utf-8") as f:
        f.write("\n".join(lines))

File: packages/cz-cli/src/commands/workspace.ts
import type { Argv } from "yargs"
import { commandGroup } from "../command-group.js"
import { JobStatus } from "@clickzetta/sdk"
import type { GlobalArgs } from "../cli.js"
import { success, error } from "../output/index.js"
import { loadProfiles, saveProfiles } from "../connection/profile-store.js"
import * as Profile from "../connection/profile-context.js"
import { logOperation } from "../logger.js"
import { getExecContext, execSql, isQueryResult, classifyExecError } from "./exec.js"

export function registerWorkspaceCommand(cli: Argv<GlobalArgs>): void {
  cli.command("workspace", "Manage workspace", (yargs) => {
    yargs
      .command(
        "list",
        "List available workspaces",
        () => {},
        async (argv) => {
          const format = argv.format
          try {
            const ctx = await getExecContext(argv)
            const sql = "SHOW WORKSPACES"
            const t0 = Date.now()
            const r = await execSql(ctx, sql)
            if (!isQueryResult(r) || r.status === JobStatus.FAILED) {
              const msg = isQueryResult(r) ? (r.errorMessage ?? "Query failed") : "Unexpected result"
              error(isQueryResult(r) ? (r.errorCode ?? "SQL_ERROR") : "SQL_ERROR", msg, { format }); return
            }
            const workspaces = r.rows.map((row) => row[0])
            success(workspaces, { format, timeMs: Date.now() - t0 })
          } catch (err) {
            const { code: _ec, message: _em, aiMessage: _ea } = classifyExecError(err)
            error(_ec, _em, { format, ...(_ea && { aiMessage: _ea }) })
          }
        },
      )
      .command(
        "current",
        "Show current workspace",
        () => {},
        async (argv) => {
          const format = argv.format
          try {
            const ctx = await getExecContext(argv)
            const sql = "SELECT current_workspace()"
            const t0 = Date.now()
            const r = await execSql(ctx, sql)
            if (!isQueryResult(r) || r.status === JobStatus.FAILED) {
              const msg = isQueryResult(r) ? (r.errorMessage ?? "Query failed") : "Unexpected result"
              logOperation("workspace current", { sql, ok: false, timeMs: Date.now() - t0 })
              error(isQueryResult(r) ? (r.errorCode ?? "SQL_ERROR") : "SQL_ERROR", msg, { format }); return
            }
            const ws = r.rows[0] ? r.rows[0][0] : null
            if (!ws) {
              logOperation("workspace current", { sql, ok: false, timeMs: Date.now() - t0 })
              error("NO_RESULT", "No current workspace set. Use `cz-cli workspace use <name>` to set one.", {
                format,
                aiMessage: "No workspace is active. List available workspaces with: cz-cli workspace list, then set one with: cz-cli workspace use <name>",
              })
              return
            }
            logOperation("workspace current", { sql, ok: true, timeMs: Date.now() - t0 })
            success({ workspace: ws }, { format, timeMs: Date.now() - t0 })
          } catch (err) {
            const { code: _ec, message: _em, aiMessage: _ea } = classifyExecError(err)
            error(_ec, _em, { format, ...(_ea && { aiMessage: _ea }) })
          }
        },
      )
      .command(
        "use <name>",
        "Switch workspace (use --persist to save to profile)",
        (y) =>
          y
            .positional("name", { type: "string", demandOption: true, describe: "Workspace name" })
            .option("schema", { type: "string", describe: "Default schema to set alongside workspace" })
            .option("persist", { type: "boolean", default: true, describe: "Save workspace to profile config (permanent). Use --no-persist to only show the SDK hint without saving." }),
        async (argv) => {
          const format = argv.format
          const name = argv.name as string
          const schemaVal = argv.schema as string | undefined
          if (argv.persist) {
            try {
              const profiles = loadProfiles()
              // Profile.current() (CZ_PROFILE, falling back to profiles.toml's
              // default_profile) is the single semantic source for "which profile is
              // active" — see its own docstring. The old fallback here
              // (getDefaultProfileName() ?? Object.keys(profiles)[0]) skipped
              // CZ_PROFILE entirely, so a `-p other` invocation with --no-persist-arg
              // could WRITE workspace/schema into a different profile than the one
              // the rest of the command actually queried against. Falling further to
              // "the first profile in the file" when even the default is unset is
              // worse here than on a read path: this persists a mutation into
              // profiles.toml, and Profile.current()'s own contract is that an
              // undefined result means "no profile configured", not "guess one".
              const profileName = argv.profile ?? Profile.current()
              if (!profileName) {
                error("PROFILE_NOT_FOUND", "No profile is active. Pass -p <profile> or set default_profile in profiles.toml.", { format })
                return
              }
              if (!profiles[profileName]) {
                error("PROFILE_NOT_FOUND", `Profile '${profileName}' not found. Create a profile first.`, { format })
                return
              }
              profiles[profileName].workspace = name
              if (schemaVal) profiles[profileName].schema = schemaVal
              saveProfiles(profiles)
              logOperation("workspace use", { ok: true })
              success(
                {
                  message: `Switched to workspace '${name}' and updated profile '${profileName}'`,
                  workspace: name,
                  schema: schemaVal ?? (profiles[profileName].schema as string | undefined) ?? "public",
                },
                { format },
              )
            } catch (err) {
              const { code: _ec, message: _em, aiMessage: _ea } = classifyExecError(err)
              error(_ec, _em, { format, ...(_ea && { aiMessage: _ea }) })
            }
          } else {
            const schemaName = schemaVal ?? "public"
            logOperation("workspace use", { ok: true })
            success(
              {
                message: `To use workspace '${name}', set SDK hint: {"sdk.job.default.ns": "${name}.${schemaName}"}`,
                workspace: name,
                schema: schemaName,
                note: "Use --persist to save this to your profile configuration",
              },
              { format },
            )
          }
        },
      )
    return commandGroup(yargs, "workspace")
  })
}
